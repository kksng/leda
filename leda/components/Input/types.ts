import * as React from 'react';
import { COMPONENTS_NAMESPACES } from '../../constants';
import { PartialGlobalDefaultTheme } from '../../utils/useTheme';
import { ValidationProps } from '../Validation/types';
import { predefinedAllowedSymbols, predefinedForbiddenSymbols } from './constants';
import { CustomRender } from '../../commonTypes';
import { DivProps } from '../Div';

export type PredefinedAllowedSymbols = keyof typeof predefinedAllowedSymbols;

export type PredefinedForbiddenSymbols = keyof typeof predefinedForbiddenSymbols;

export interface ClearEvent extends React.MouseEvent<HTMLInputElement> {
  component: {
    value: string,
    name?: string,
  },
}

export interface TypeEvent extends React.ChangeEvent<HTMLInputElement> {
  component: {
    value: string,
    name?: string,
  },
}

export interface ResetEvent {
  currentTarget?: undefined,
  component: {
    value: string,
    name?: string,
  },
}

export interface EnterPressEvent extends React.KeyboardEvent<HTMLInputElement> {
  component: {
    value: string,
    name?: string,
  },
}

export type ChangeEvent = TypeEvent | ClearEvent | ResetEvent;

export interface BlurEvent extends React.FocusEvent<HTMLInputElement> {
  component: {
    value: string,
    name?: string,
    isValid: boolean,
  },
}

export interface FocusEvent extends React.FocusEvent<HTMLInputElement> {
  component: {
    value: string,
    name?: string,
  },
}

export interface InputProps extends ValidationProps {
  /** Позволяет вводить в инпут только символы, удовлеторвяющие RegExp или из списка предопределённых */
  allowedSymbols?: PredefinedAllowedSymbols | RegExp,
  /** Значение по умолчанию */
  defaultValue?: string | null,
  /** Запрещает вводить в инпут символы, удовлеторвяющие RegExp или из списка предопределённых */
  forbiddenSymbols?: PredefinedForbiddenSymbols | RegExp,
  /** Отображение кнопки очистки в инпуте */
  hasClearButton?: boolean,
  /** Отключенное состояние инпута */
  isDisabled?: boolean,
  /** Рендер инпута */
  inputRender?: CustomRender<InputProps, InputState, React.InputHTMLAttributes<HTMLInputElement>>,
  /** Переводит все вводимые буквы в верхний или нижний регистр */
  letterCase?: 'lower' | 'upper',
  /** Максимальная длина введенного значения */
  maxLength?: number,
  /** Имя компонента, используется в валидации и для сохранения данных из формы */
  name?: string,
  /** Обработчик блюра */
  onBlur?: (ev: BlurEvent) => void,
  /** Обработчик изменения */
  onChange?: (ev: ChangeEvent) => void,
  /** Обработчик нажатия Enter */
  onEnterPress?: (ev: EnterPressEvent) => void,
  /** Обработчик фокуса */
  onFocus?: (ev: FocusEvent) => void,
  /** Реф */
  ref?: React.Ref<InputRefCurrent>,
  /** Тема компонента */
  theme?: PartialGlobalDefaultTheme[typeof COMPONENTS_NAMESPACES.input],
  /** Значение для инпута */
  value?: string | null,
  /** Рендер враппера */
  wrapperRender?: CustomRender<InputProps, InputState, DivProps>,
  /** Классы переданные через _ */
  [x: string]: unknown,
}

export interface InputState {
  isFocused: boolean,
  isValid: boolean,
  value: string,
}
export interface InputRefCurrent {
  wrapper: HTMLDivElement | null,
  input: HTMLInputElement | null,
}
