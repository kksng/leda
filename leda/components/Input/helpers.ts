import { isString } from 'lodash';
import { predefinedAllowedSymbols, predefinedForbiddenSymbols } from './constants';
import { InputProps } from './types';

export const isSymbolAllowed = (value: string, allowedSymbols?: keyof typeof predefinedAllowedSymbols | RegExp): boolean => {
  if (!allowedSymbols || value.length === 0) return true;

  if (isString(allowedSymbols)) {
    const predefinedRegExp = predefinedAllowedSymbols[allowedSymbols];

    if (!predefinedRegExp) throw new Error(`L.Input: no such predefined allowedSymbols - "${allowedSymbols}"!`);

    return [...value].every((symbol) => predefinedRegExp.test(symbol));
  }

  if (allowedSymbols instanceof RegExp) {
    const regExp: RegExp = allowedSymbols;
    return [...value].every((symbol) => regExp.test(symbol));
  }

  throw new Error('L.Input: allowedSymbols prop accepts only predefined string or RegExp!');
};


export const isSymbolForbidden = (value: string, forbiddenSymbols?: keyof typeof predefinedForbiddenSymbols | RegExp): boolean => {
  if (!forbiddenSymbols || value.length === 0) return false;

  if (isString(forbiddenSymbols)) {
    const predefinedRegExp = predefinedForbiddenSymbols[forbiddenSymbols];

    if (!predefinedRegExp) throw new Error(`L.Input: no such predefined forbiddenSymbols - "${forbiddenSymbols}"!`);

    return [...value].some((symbol) => predefinedRegExp.test(symbol));
  }

  if (forbiddenSymbols instanceof RegExp) {
    const regExp: RegExp = forbiddenSymbols;
    return [...value].some((symbol) => regExp.test(symbol));
  }

  throw new Error('L.Input: forbiddenSymbols prop accepts only predefined string or RegExp!');
};

export const transformToCase = (letter: string, letterCase: InputProps['letterCase']): string | never => {
  if (letterCase === 'lower') return letter.toLowerCase();

  if (letterCase === 'upper') return letter.toUpperCase();

  throw new Error('L.Input: letterCase prop must be \'lower\' or \'upper\'!');
};

export const getValue = (valueProp: string | null | undefined, valueState: string): string => {
  if (valueProp === undefined) return valueState;

  if (valueProp === null) {
    return '';
  }

  return valueProp;
};
