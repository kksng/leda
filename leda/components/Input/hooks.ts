import * as React from 'react';
import { SetState } from '../../commonTypes';
import { InputProps } from './types';

export const useReset = (
  props: InputProps,
  setValue: SetState<string>,
) => React.useCallback(() => {
  const newValue = props.defaultValue || '';

  setValue(newValue);

  if (props.onChange) {
    const newEvent = {
      component: {
        name: props.name,
        value: newValue,
      },
    };

    props.onChange(newEvent);
  }
}, [props, setValue]);
