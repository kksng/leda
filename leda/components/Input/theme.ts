/* eslint-disable key-spacing */
export const defaultInputTheme = {
  closeIcon:                'input-clear-icon',
  icon:                     'input-icon',
  iconLeft:                 'icon-left',
  iconRight:                'icon-right',
  input:                    'input-element',
  inputWrapper:             'input-element-wrapper',
  inputWrapperDisabled:     'disabled',
  inputWrapperFocused:      'focused',
  inputWrapperInvalid:      'danger',
  inputWrapperRequired:     'required',
  prefix:                   'input-prefix',
  suffix:                   'input-suffix',
  wrapper:                  'input-wrapper',
};
/* eslint-enable key-spacing */
