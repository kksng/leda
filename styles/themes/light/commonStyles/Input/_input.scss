// общие стили для инпутов

%input {
  outline: none;
  color: $black-01;
  border: 1px solid $black-05;
  background-color: $white;
  padding: 5px 4px 7px 12px;
  border-radius: $border-radius;
  display: flex;
  justify-content: space-between;
  height: 32px;
  font-size: 14px;
}

%input-wrapper {
  color: $black-01;
  align-items: center;
  border: 1px solid $black-05;
  background-color: $white;
  border-radius: $border-radius;
  display: flex;
  justify-content: space-between;
  height: 32px;

  & > input[class*="input"] {
    background-color: transparent;
    border: none;
    outline: none;
    width: 100%;
    height: 30px;
    border-radius: $border-radius;
    font-family: inherit;
    font-size: 14px;
    padding: 0 4px 0 8px;
    min-width: 0;

    &::placeholder {
      @extend %input-placeholder;
    }
  }
}

%input-hovered {
  border-color: $black-02;
}

%input-focused {
  border: 1px solid $black-02;
  box-shadow: none;
}

%input-disabled {
  border-color: $black-05;
  // Из-за important в styles-esphere
  background-color: $black-07 !important;
  color: $black-03 !important;
}

%input-placeholder {
  color: $black-03;
}

%input-prefix {
  margin-left: 8px;
}

%input-suffix {
  margin-right: 14px;
}

%input-prefix, %input-suffix {
  user-select: none;
  flex: none;
  white-space: nowrap;
  color: $black-03;
}

%input-icon-search {
  &:before {
    content: '\e92c';
    color:  $black-03;
    width: 20px;
    height: 20px;
    display: block;
  }
}
