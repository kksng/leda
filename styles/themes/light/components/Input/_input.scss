.input {
  &-wrapper {
    display: flex;
    flex-direction: column;
    position: relative;
  }

  &-element-wrapper {
    @extend %input-wrapper;
    padding: 0;

    &:not(.disabled):not(.danger):hover {
      @extend %input-hovered;
    }

    &:not(.focused):not(.disabled).danger {
      border-color: $error-02;
    }

    &.focused {
      @extend %input-focused;
    }

    &.disabled {
      @extend %input-disabled;
    }
  }

  &-clear-icon {
    font-size: 18px;
    width: 30px;
    height: 30px;
    cursor: pointer;
    position: absolute;
    right: 0;

    &::before {
      font-family: "Font Awesome 5 Free";
      content: "\00D7";
      color: $black-03;
      position: absolute;
      top: 45%;
      left: 50%;
      transform: translate(-50%, -50%);
      display: block;
      font-style: normal;
    }
  }

  &-icon {
    margin: 0;
    font-size: 14px;
    color: $black-03;

    &:before {
      position: relative;
      top: 5px;
    }

    &.icon-left {
      padding-left: 5px;

      & + input {
        padding-left: 7px;
      }
    }

    &.icon-right {
      padding-right: 7px;
    }
  }

  &-prefix {
    @extend %input-prefix;
  }

  &-suffix {
    @extend %input-suffix;
  }
}
