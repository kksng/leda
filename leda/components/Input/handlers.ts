import * as React from 'react';
import { CustomEventHandler, SetState } from '../../commonTypes';
import { InputProps } from './types';
import { isSymbolAllowed, isSymbolForbidden, transformToCase } from './helpers';

export const createChangeHandler = (
  props: InputProps,
  setValue: SetState<string>,
): CustomEventHandler<React.ChangeEvent<HTMLInputElement>> => (event) => {
  if (props.maxLength && props.maxLength < event.target.value.length) return;

  if (isSymbolForbidden(event.target.value, props.forbiddenSymbols)) return;

  if (!isSymbolAllowed(event.target.value, props.allowedSymbols)) return;

  const newValue = props.letterCase ? transformToCase(event.target.value, props.letterCase) : event.target.value;

  if (props.value === undefined) {
    setValue(newValue);
  }

  props.onChange?.({
    ...event,
    component: {
      value: newValue,
      name: props.name,
    },
  });
};

export const createClearHandler = (
  props: InputProps,
  setValue: SetState<string>,
): CustomEventHandler<React.MouseEvent<HTMLInputElement>> => (event) => {
  event.preventDefault();

  if (props.value === undefined) {
    setValue('');
  }

  props.onChange?.({
    ...event,
    component: {
      value: '',
      name: props.name,
    },
  });
};

export const createBlurHandler = (
  props: InputProps,
  setFocused: SetState<boolean>,
  validate: () => boolean,
): React.FocusEventHandler<HTMLInputElement> => (event) => {
  setFocused(false);

  const newValid = validate();

  props.onBlur?.({
    ...event,
    component: {
      value: event.target.value,
      name: props.name,
      isValid: newValid,
    },
  });
};

export const createFocusHandler = (
  props: InputProps,
  isValid: boolean,
  setFocused: SetState<boolean>,
): React.FocusEventHandler<HTMLInputElement> => (event) => {
  setFocused(true);

  props.onFocus?.({
    ...event,
    component: {
      value: event.target.value,
      name: props.name,
      isValid,
    },
  });
};

export const createKeyDownHandler = (
  props: InputProps,
): React.KeyboardEventHandler<HTMLInputElement> => (event) => {
  if (event.key === 'Enter') {
    props.onEnterPress?.({
      ...event,
      component: {
        value: event.currentTarget.value,
        name: props.name,
      },
    });
  }
};
