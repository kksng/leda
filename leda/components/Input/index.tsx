import React from 'react';
import { COMPONENTS_NAMESPACES } from '../../constants';
import {
  mergeClassNames, getClassNames, bindFunctionalRef, useTheme, useElement,
} from '../../utils';
import { useValidation } from '../Validation';
import { InputProps, InputRefCurrent } from './types';
import { Div } from '../Div';
import {
  createBlurHandler,
  createChangeHandler,
  createClearHandler,
  createFocusHandler,
  createKeyDownHandler,
  createResetHandler,
} from './handlers';
import { getValue } from './helpers';

// как валидировать инпуты: ../Validation/validation.md

export const Input = React.forwardRef((props: InputProps, ref: React.Ref<InputRefCurrent>): React.ReactElement => {
  const {
    allowedSymbols,
    className,
    defaultValue,
    forbiddenSymbols,
    form,
    hasClearButton,
    inputRender,
    invalidMessage,
    isDisabled,
    isRequired,
    isValid: isValidProp,
    letterCase,
    maxLength,
    name,
    onBlur, // exclude from restProps
    onChange,
    onEnterPress,
    onFocus,
    requiredMessage,
    shouldValidateUnmounted,
    theme: themeProp,
    validationMessageRender,
    validator,
    value: valueProp,
    wrapperRender,
    invalidMessageRender,
    ...restProps
  } = mergeClassNames<InputProps>(props);

  const [value, setValue] = React.useState<string>(defaultValue || '');

  const {
    isValid, validateCurrent, InvalidMessage,
  } = useValidation(props, {
    value,
  }, {
    reset: createResetHandler({
      props, setValue, value: defaultValue || '',
    }),
  });

  const theme = useTheme(themeProp, COMPONENTS_NAMESPACES.input);

  const [isFocused, setFocused] = React.useState(false);

  const state = { value, isFocused, isValid };

  const handleBlur = createBlurHandler(props, state, setFocused, validateCurrent);

  const handleChange = createChangeHandler(props, state, setValue);

  const handleFocus = createFocusHandler(props, state, setFocused);

  const handleClearValue = createClearHandler(props, state, setValue);

  const handleKeyDown = createKeyDownHandler(props);

  const wrapperClassNames = getClassNames(
    className,
    theme.wrapper,
  );

  const inputClassNames = getClassNames(
    theme.inputWrapper,
    {
      [theme.inputWrapperFocused]: isFocused,
      [theme.inputWrapperDisabled]: isDisabled,
      [theme.inputWrapperInvalid]: !isValid,
    },
  );

  // todo: fix hasClearButton in controlled mode
  const shouldRenderClearButton = value && value.length > 0 && hasClearButton;

  const Wrapper = useElement(
    'Wrapper',
    Div,
    wrapperRender,
    props,
    state,
  );

  const InputElement = useElement(
    'Input',
    'input' as unknown as React.FC<React.InputHTMLAttributes<HTMLInputElement>>,
    inputRender,
    props,
    state,
  );

  return (
    <Wrapper
      className={wrapperClassNames}
      ref={ref && ((component) => bindFunctionalRef(component, ref, component && component.wrapper && {
        wrapper: component.wrapper,
        input: component.wrapper.querySelector('input'),
      }))}
    >
      <Div
        className={inputClassNames}
      >
        <InputElement
          {...restProps}
          aria-invalid={!isValid}
          aria-required={isRequired}
          className={theme.input}
          disabled={isDisabled}
          form={form}
          name={name}
          onBlur={handleBlur}
          onChange={handleChange}
          onFocus={handleFocus}
          onKeyDown={handleKeyDown}
          value={getValue(valueProp, value)}
        />
        {
          shouldRenderClearButton && (
            <i
              className={theme.closeIcon}
              onMouseDown={handleClearValue}
            />
          )
        }
      </Div>
      {!isFocused && !isDisabled && (
        <InvalidMessage />
      )}
    </Wrapper>
  );
}) as React.FC<InputProps>;

Input.displayName = 'Input';
