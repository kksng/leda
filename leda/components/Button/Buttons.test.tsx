// @ts-nocheck
import React from 'react';
import {
  render,
  fireEvent,
} from '@testing-library/react';
import { Button } from './index';
import { Input } from '../Input';


describe('Check Button snapshots collection', () => {
  test('is Button with basic pros render right?', () => {
    const wrapper = render((
      <Button>test</Button>
    ));
    expect(wrapper.container)
      .toMatchSnapshot();
  });
  test('is Button with isLoading attributes render right?', () => {
    const wrapper = render((
      <Button isLoading>test</Button>
    ));
    expect(wrapper.container)
      .toMatchSnapshot();
  });
  test('is Button with isDisabled attributes render right?', () => {
    const wrapper = render((
      <Button isDisabled>test</Button>
    ));
    expect(wrapper.container)
      .toMatchSnapshot();
  });
  test('is Button with Form and onClick props render right?', () => {
    const wrapper = render((
      <div>
        <Input form="test" name="in" isRequired />
        <Button onClick={jest.fn()} form="test" isDisabled>test</Button>
      </div>
    ));
    expect(wrapper.container)
      .toMatchSnapshot();
  });
});
describe('Check Button condition set collection', () => {
  test('is Button attributes set right?', () => {
    const { container, rerender, debug } = render(<Button _primary>test</Button>);

    expect(container.querySelectorAll('button.primary')[0])
      .toBeInTheDocument();

    rerender(<Button _secondary>test</Button>);

    expect(container.querySelectorAll('button.secondary')[0])
      .toBeInTheDocument();

    rerender(<Button _success>test</Button>);

    expect(container.querySelectorAll('button.success')[0])
      .toBeInTheDocument();

    rerender(<Button _warning>test</Button>);

    expect(container.querySelectorAll('button.warning')[0])
      .toBeInTheDocument();

    rerender(<Button _danger>test</Button>);

    expect(container.querySelectorAll('button.danger')[0])
      .toBeInTheDocument();

    rerender(<Button _small>test</Button>);

    expect(container.querySelectorAll('button.small')[0])
      .toBeInTheDocument();

    rerender(<Button _large>test</Button>);

    expect(container.querySelectorAll('button.large')[0])
      .toBeInTheDocument();

    rerender(<Button _block>test</Button>);

    expect(container.querySelectorAll('button.block')[0])
      .toBeInTheDocument();
  });
  test('is Button isDisabled property work right?', () => {
    const onClickHandler = jest.fn();
    const { container, debug } = render(<Button isDisabled onClick={onClickHandler}>test</Button>);
    const button = container.querySelectorAll('button')[0];

    expect(container.querySelectorAll('button.disabled')[0])
      .toBeInTheDocument();

    fireEvent.click(button);

    expect(onClickHandler)
      .not
      .toHaveBeenCalled();
  });
  test('is Button isLoading property work right?', () => {
    const onClickHandler = jest.fn();
    const { container } = render(<Button isLoading onClick={onClickHandler}>test</Button>);
    const button = container.querySelectorAll('button')[0];

    expect(container.querySelectorAll('button.loading')[0])
      .toBeInTheDocument();

    fireEvent.click(button);

    expect(onClickHandler)
      .not
      .toHaveBeenCalled();
  });
});
describe('Check Button form validation test collection', () => {
  test('is Button with form attributes submit Form with valid value right?', () => {
    const onSubmitValid = jest.fn();
    const onSubmitInvalid = jest.fn();
    const email = 'some@email.ru';
    const { getByText } = render((
      <div>
        <Input form="test" name="email" invalidMessage="error" isRequired validator="email" value={email} />
        <Button onClick={onSubmitValid} onValidationFail={onSubmitInvalid} form="test">test</Button>
      </div>
    ));

    fireEvent.click(getByText('test'));
    expect(onSubmitValid).toBeCalledTimes(1);
    expect(onSubmitInvalid).toBeCalledTimes(0);
  });
  test('is Button with form attributes submit Form with invalid value right?', () => {
    const onSubmitValid = jest.fn();
    const onSubmitInvalid = jest.fn();
    const email = '-----';
    const { getByText } = render((
      <div>
        <Input form="test" name="email" invalidMessage="error" isRequired validator="email" value={email} />
        <Button onClick={onSubmitValid} onValidationFail={onSubmitInvalid} form="test">test</Button>
      </div>
    ));

    fireEvent.click(getByText('test'));
    expect(onSubmitValid).toBeCalledTimes(0);
    expect(onSubmitInvalid).toBeCalledTimes(1);
  });
});
