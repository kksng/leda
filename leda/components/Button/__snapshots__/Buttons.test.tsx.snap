// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check Button snapshots collection is Button with Form and onClick props render right? 1`] = `
<div>
  <div>
    <div
      class="input-wrapper"
    >
      <div
        class="input-element-wrapper required"
      >
        <input
          aria-invalid="false"
          aria-required="true"
          class="input-element"
          form="test"
          name="in"
          value=""
        />
      </div>
    </div>
    <button
      class="button-wrapper disabled"
      type="button"
    >
      test
    </button>
  </div>
</div>
`;

exports[`Check Button snapshots collection is Button with basic pros render right? 1`] = `
<div>
  <button
    class="button-wrapper"
    type="button"
  >
    test
  </button>
</div>
`;

exports[`Check Button snapshots collection is Button with isDisabled attributes render right? 1`] = `
<div>
  <button
    class="button-wrapper disabled"
    type="button"
  >
    test
  </button>
</div>
`;

exports[`Check Button snapshots collection is Button with isLoading attributes render right? 1`] = `
<div>
  <button
    class="button-wrapper loading"
    type="button"
  >
    test
  </button>
</div>
`;
